### ISSUE TITLE
New Logger Interface

### ISSUE THREAD

**Issue by dhui** [golang-migrate/migrate#364]
## Problem
The current logger interface is simple but doesn't support richer features.

## Desired features
1. Log level
1. Log sinks
    * See: https://github.com/golang-migrate/migrate/issues/363
1. Configurable

**Comment by dhui** [comment]
https://github.com/golang-migrate/migrate/issues/363#issuecomment-605487428

> What do you think about using Zap logger? It's fast, configurable, stable. I believe that it satisfy all the requirements, you wrote in #364. If you think it's a good idea, I could help with its integration.

@renta Using `zap` or `zerolog` would be fine. But instead of using them directly, I'd wrap them and use our own logger interface.

**Comment by tarungulati1988** [comment]
I'd like to work on this

**Comment by dhui** [comment]
@tarungulati1988 any help is appreciated! 🙏

**Comment by milhamsuryapratama** [comment]
```
2025-07-19T06:36:47+07:00 [INFO] 1/u create_users (10.637666ms)
2025-07-19T06:36:47+07:00 [INFO] 2/u add_email (19.422958ms)
```
is the expected output like this, right ? @dhui @dzoeteman @SimonWoidig

**Comment by milhamsuryapratama** [comment]
@SimonWoidig as I know,`log/slog` is available for golang 1.21 version. when you use `log/slog`, is that will working on golang under 1.21 version ?

**Comment by SimonWoidig** [comment]
@milhamsuryapratama 
Good point. But if you just declare an interface, you shouldn't worry about slog being or not being in the go version.
Another problem arises for applications running on 1.21 or lower - since there is no slog, the users would have to write their own implementation. This I don't know how to solve correctly. Maybe one of the maintainers can give a good solution.

**Comment by SimonWoidig** [comment]
@SimonWoidig Well that doesn't really work with slog. Slog doesn't really use the `InfoF`, `ErrorF` and such other formatting functions. It instead uses attributes. But I don't know, how to integrate this in the older Go versions.
